cpp-client/deephaven/dhcore/src/utility/utility.cc

```diff
@@ -187,5 +196,59 @@ void dumpTillPercentOrEnd(ostream &result, const char **fmt) {
   result.write(start, p - start);
   *fmt = p;
 }
+
+const char DIGITS[] = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' };
+static_assert(sizeof(DIGITS) == 10);
 }  // namespace
+
+std::string
+EpochMillisToStr(std::int64_t epochMillis) {
+  const time_t secs = epochMillis / 1000;
+  const unsigned millisRest = epochMillis % 1000;
+  struct tm tm;
+  localtime_r(&secs, &tm);
+  std::size_t resultSz = 32u;  // we need 28u + 1u; +1 for null terminated.
+  std::string result(resultSz, '#');
+  char *resBuf = &result[0];
+  strftime(resBuf, resultSz,
+           "%F"    // 10 chars: ISO-8601 date format: `YYYY-mm-dd`
+           "T"     //  1 char:  A literal "T" separator (ISO-8601).
+           "%T"    //  8 chars: ISO-8601 time format: `HH:MM:SS`
+           ".000"  //  4 chars: Placeholder for msecs
+           "%z",   //  3 chars: timezone [+-]hhmm
+           &tm);
+  const unsigned hundredMillisRest = millisRest / 100;
+  const unsigned millisRestMod100 = millisRest % 100;
+  const unsigned tensMillisRest = millisRestMod100 / 10;
+  const unsigned singleMillisRest = millisRestMod100 % 10;
+
+  assert(hundredMillisRest < 10);
+  assert(resBuf[20] == '0');
+  resBuf[20] = DIGITS[hundredMillisRest];
+  assert(tensMillisRest < 10);
+  assert(resBuf[21] == '0');
+  resBuf[21] = DIGITS[tensMillisRest];
+  assert(singleMillisRest < 10);
+  assert(resBuf[22] == '0');
+  resBuf[22] = DIGITS[singleMillisRest];
+  return result;
+}
```

Comment:
This is very subjective, but this is a little too much array surgery for my tastes. I have a method in `dhcore/src/types.cc` that does almost the same thing (but it prints nanos rather than millis, it prints in UTC rather than localtime, and it writes to a stream rather than returning a string). I think I would request that you use that approach (modifed of course for your purposes). You could use a SimpleOstringStream for example. That method looks like the below. Subjective, but I sort of prefer making little local C arrays and then stitching them together at the end, rather than surgery in-place.

```
void DateTime::streamIrisRepresentation(std::ostream &s) const {
  size_t oneBillion = 1000000000;
  time_t timeSecs = nanos_ / oneBillion;
  auto nanos = nanos_ % oneBillion;
  struct tm tm = {};
  gmtime_r(&timeSecs, &tm);
  char dateBuffer[32];  // ample
  char nanosBuffer[32];  // ample
  strftime(dateBuffer, sizeof(dateBuffer), "%FT%T", &tm);
  snprintf(nanosBuffer, sizeof(nanosBuffer), "%09zd", nanos);
  s << dateBuffer << '.' << nanosBuffer << " UTC";
}
```

Reply:
We discussed a bit over slack.  I am punting on doing something right now, in light of there being no correctness objection, and the amount of work associated with this one is relatively high (compared to addressing all the rest of the comments).  So my proposal is we move forward with this implementation for merge, and if we care enough we address this in a followup PR later.

Reply:
I now have a correctness objection, because both epochMillisToStr() and objectId() return garbled strings. (I mentioned the problem with objectId() in a previous comment -- maybe you didn't see it?)

`objectId()` is incorrect because it makes output like
```
/home/kosak/git/deephaven-core/cpp-client/tests/utility_test.cc:29: Failure:
  CHECK(id == "hello[0xdeadbeef]")
with expansion:
  "hello[0x0xdeadbeef\x0000000]"
==
"hello[0xdeadbeef]"
```

with repeated "0x0x", an embedded \0, and other problems.

`epochMillisToStr()` is incorrect because it makes output like
```
/home/kosak/git/deephaven-core/cpp-client/tests/utility_test.cc:22: Failure:
  CHECK(d1 == "2023-07-12T00:40:24.000-0400")
with expansion:
  "2023-07-12T00:40:24.000-0400\x00###"
==
"2023-07-12T00:40:24.000-0400"
```

where there is an embedded \0 after the 0400 and trailing matter with the pound signs.

I've implemented both in the way I think makes sense and I have taken the liberty of pushing two commits plus simple unit tests to your branch, Kindly take a look. The second commit undoes a mistake I made in the first, where I accidentally trashed another unrelated unit test.

The unit test for epochMillisToStr() is not going to really work on the server because the server may have a different notion of local time. We may have to do something about that -- either not committing the unit test, forcing the server to temporarily believe in a certain local time zone, or some other idea.